File: server/routes/health.ts
import type { Hono } from 'hono'
import { openApiDocument } from '../openapi'
import { db, tursoUrl } from '../db'
import { getIceServers } from '../turn'
import { inc, prometheusText, snapshot } from '../metrics'
import { rateLimit } from '../rateLimit'
import { requireAdmin } from '../security'
import { clientIp, resolveVersion } from '../http'
import { queueStats } from '../matchmaking'
import { config } from '../config'
import {
  API_ROUTES,
  HTTP_HEADERS,
  HTTP_STATUS,
  MIME_TYPE,
  WS_PATH,
  API_PREFIX,
} from '../../shared/constants'

const APP_VERSION = resolveVersion()

export function registerHealthRoutes(
  app: Hono,
  appUrl: string,
  isDraining?: () => boolean,
  dbOk?: { value: boolean },
) {
  app.get(API_ROUTES.docs, (c) => c.json(openApiDocument(appUrl)))

  app.get(API_ROUTES.health, (c) => {
    const stats = queueStats()
    return c.json({
      ok: !(isDraining?.() ?? false) && (dbOk?.value ?? true),
      version: APP_VERSION,
      draining: isDraining?.() ?? false,
      waiting: stats.waiting,
      online: stats.online,
      database: tursoUrl.startsWith('file:') ? 'local libSQL' : 'turso',
      turn: Boolean(process.env.TURN_SECRET && process.env.TURN_URLS),
      uptimeSec: Math.floor(process.uptime()),
      features: config.features,
    })
  })

  app.get(API_ROUTES.healthLive, (c) => c.json({ ok: true, version: APP_VERSION }))

  app.get(API_ROUTES.healthReady, async (c) => {
    if (isDraining?.()) return c.json({ ok: false, reason: 'draining' }, HTTP_STATUS.serviceUnavailable)
    try {
      await db.execute('SELECT 1')
      if (dbOk) dbOk.value = true
    } catch {
      if (dbOk) dbOk.value = false
      return c.json({ ok: false, reason: 'database' }, HTTP_STATUS.serviceUnavailable)
    }
    return c.json({ ok: true })
  })

  app.get(API_ROUTES.metrics, (c) => {
    if (!config.metricsPublic && !requireAdmin(c)) {
      return c.json({ error: 'Forbidden' }, HTTP_STATUS.forbidden)
    }
    const stats = queueStats()
    return c.json({ ...snapshot(), queue: stats, draining: isDraining?.() ?? false })
  })

  app.get(API_ROUTES.metricsPrometheus, (c) => {
    if (!config.metricsPublic && !requireAdmin(c)) {
      return c.text('Forbidden', HTTP_STATUS.forbidden)
    }
    const stats = queueStats()
    const body = prometheusText({
      queue_waiting: stats.waiting,
      queue_online: stats.online,
      draining: isDraining?.() ? 1 : 0,
    })
    return c.body(body, HTTP_STATUS.ok, { [HTTP_HEADERS.contentType]: MIME_TYPE.prometheus })
  })

  app.get(API_ROUTES.configPublic, (c) =>
    c.json({
      features: {
        anonymousMatch: config.features.anonymousMatch,
        qualityTelemetry: config.features.qualityTelemetry,
      },
      turnConfigured: Boolean(process.env.TURN_SECRET && process.env.TURN_URLS),
    }),
  )

  app.get(API_ROUTES.ice, async (c) => {
    const ip = clientIp(c)
    if (!rateLimit(`ice:${ip}`, 30, 60_000)) return c.json({ error: 'Too many requests' }, HTTP_STATUS.tooManyRequests)
    return c.json(getIceServers())
  })
}

File: rust/src/routes/health.rs
//! Health, metrics, public config and ICE. Port of `server/routes/health.ts`.

use axum::extract::State;
use axum::http::{HeaderMap, StatusCode};
use axum::response::{IntoResponse, Response};
use axum::routing::get;
use axum::{Json, Router};
use serde_json::{json, Value};

use crate::error::{ApiError, ApiResult};
use crate::infra::http::client_ip;
use crate::infra::metrics::{prometheus_text, snapshot, uptime_sec};
use crate::infra::rate_limit::rate_limit;
use crate::infra::security::require_admin;
use crate::infra::version::app_version;
use crate::turn::{ice_servers, turn_configured};
use crate::AppState;

pub fn router(state: AppState) -> Router {
    Router::new()
        .route("/api/v1/docs", get(docs))
        .route("/api/v1/health", get(health))
        .route("/api/v1/health/live", get(live))
        .route("/api/v1/health/ready", get(ready))
        .route("/api/v1/metrics", get(metrics))
        .route("/api/v1/metrics/prometheus", get(metrics_prometheus))
        .route("/api/v1/config/public", get(config_public))
        .route("/api/v1/ice", get(ice))
        .with_state(state)
}

async fn docs(State(state): State<AppState>) -> Json<Value> {
    Json(crate::openapi::open_api_document(&state.config.app_url))
}

async fn live() -> Json<Value> {
    Json(json!({ "ok": true, "version": app_version() }))
}

async fn health(State(state): State<AppState>) -> Json<Value> {
    let stats = state.engine.queue_stats().await;
    let draining = state.is_draining();
    Json(json!({
        "ok": !draining && state.db_ok(),
        "version": app_version(),
        "draining": draining,
        "waiting": stats.waiting,
        "online": stats.online,
        "database": state.db.kind(),
        "turn": turn_configured(),
        "uptimeSec": uptime_sec(),
        "features": {
            "anonymousMatch": state.config.features.anonymous_match,
            "guestReports": state.config.features.guest_reports,
            "qualityTelemetry": state.config.features.quality_telemetry,
            "requireEmailVerified": state.config.features.require_email_verified,
        },
    }))
}

/// Readiness doubles as the database liveness probe: a failing `SELECT 1` flips
/// the flag that `/health` reports.
async fn ready(State(state): State<AppState>) -> Response {
    if state.is_draining() {
        return (
            StatusCode::SERVICE_UNAVAILABLE,
            Json(json!({ "ok": false, "reason": "draining" })),
        )
            .into_response();
    }
    match state.db.conn().query("SELECT 1", ()).await {
        Ok(_) => {
            state.set_db_ok(true);
            Json(json!({ "ok": true })).into_response()
        }
        Err(_) => {
            state.set_db_ok(false);
            (
                StatusCode::SERVICE_UNAVAILABLE,
                Json(json!({ "ok": false, "reason": "database" })),
            )
                .into_response()
        }
    }
}

async fn metrics(State(state): State<AppState>, headers: HeaderMap) -> ApiResult<Json<Value>> {
    if !state.config.metrics_public && !require_admin(&headers) {
        return Err(ApiError::forbidden("Forbidden"));
    }
    let stats = state.engine.queue_stats().await;
    let mut out = snapshot();
    out["queue"] = json!({ "waiting": stats.waiting, "online": stats.online });
    out["draining"] = json!(state.is_draining());
    Ok(Json(out))
}

async fn metrics_prometheus(State(state): State<AppState>, headers: HeaderMap) -> Response {
    if !state.config.metrics_public && !require_admin(&headers) {
        return (StatusCode::FORBIDDEN, "Forbidden").into_response();
    }
    let stats = state.engine.queue_stats().await;
    let body = prometheus_text(&[
        ("queue_waiting", stats.waiting as f64),
        ("queue_online", stats.online as f64),
        ("draining", if state.is_draining() { 1.0 } else { 0.0 }),
    ]);
    ([("content-type", "text/plain; version=0.0.4")], body).into_response()
}

async fn config_public(State(state): State<AppState>) -> Json<Value> {
    Json(json!({
        "features": {
            "anonymousMatch": state.config.features.anonymous_match,
            "qualityTelemetry": state.config.features.quality_telemetry,
        },
        "turnConfigured": turn_configured(),
        // The client only renders the Google button when the server can
        // actually complete the flow.
        "googleAuth": state.google_oauth.is_some(),
    }))
}

async fn ice(headers: HeaderMap) -> ApiResult<Json<Value>> {
    let ip = client_ip(&headers);
    if !rate_limit(&format!("ice:{ip}"), 30, 60_000) {
        return Err(ApiError::too_many("Too many requests"));
    }
    Ok(Json(ice_servers()))
}
